flink-runtime/src/test/java/org/apache/flink/runtime/util/profiler/ProfilingServiceTest.java

```diff
@@ -0,0 +1,170 @@
+        }
+    }
+
+    @Test
+    @Timeout(value = 1, unit = TimeUnit.MINUTES)
+    public void testAllProfilingMode() throws ExecutionException, InterruptedException {
+        for (ProfilingInfo.ProfilingMode mode : ProfilingInfo.ProfilingMode.values()) {
+            ProfilingInfo profilingInfo =
+                    profilingService.requestProfiling(resourceID, profilingDuration, mode).get();
+            if (isNoPermissionOrAllocateSymbol(profilingInfo)) {
+                LOG.warn(
+                        "Ignoring failed profiling instance in {} mode, which caused by no permission.",
+                        profilingInfo.getProfilingMode());
+                continue;
+            }
+            Assertions.assertEquals(
+                    ProfilingInfo.ProfilingStatus.RUNNING,
+                    profilingInfo.getStatus(),
+                    String.format(
+                            "Submitting profiling request should be succeed or no permission, but got errorMsg=%s",
+                            profilingInfo.getMessage()));
+            waitForProfilingFinished(profilingService);
+            Assertions.assertEquals(
+                    ProfilingInfo.ProfilingStatus.FINISHED,
+                    profilingInfo.getStatus(),
+                    String.format(
+                            "Profiling request should complete successful, but got errorMsg=%s",
+                            profilingInfo.getMessage()));
+        }
+
+        verifyRollingDeletion(profilingService);
+    }
+
+    private void verifyConfigsWorks(ProfilingService profilingService, Path configuredDir) {
+        Assertions.assertEquals(configuredDir.toString(), profilingService.getProfilingResultDir());
+        Assertions.assertEquals(historySizeLimit, profilingService.getHistorySizeLimit());
+    }
+
+    private void verifyRollingDeletion(ProfilingService profilingService) {
+        ArrayDeque<ProfilingInfo> profilingList =
+                profilingService.getProfilingListForTest(resourceID);
+        // Profiling History shouldn't exceed history size limit.
+        Assertions.assertEquals(historySizeLimit, profilingList.size());
+        // Profiling History files should be rolling deleted.
+        Set<String> resultFileNames = new HashSet<>();
+        File configuredDir = new File(profilingService.getProfilingResultDir());
+        for (File f : Objects.requireNonNull(configuredDir.listFiles())) {
+            resultFileNames.add(f.getName());
+        }
+        Assertions.assertEquals(profilingList.size(), resultFileNames.size());
+        for (ProfilingInfo profilingInfo : profilingList) {
+            String outputFile = profilingInfo.getOutputFile();
+            Assertions.assertTrue(resultFileNames.contains(outputFile));
+        }
+    }
+
+    private void waitForProfilingFinished(ProfilingService profilingService) {
+        while (!profilingService.getProfilingFuture().isDone()) {}
```

Comment:
It's better to add a `Thread#sleep` here.

Reply:
Corrected with 1 second sleep.